### PR TITLE
Run external test suite tests in pkgci.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/openxla/iree/issues/16372. Depends on https://github.com/nod-ai/SHARK-TestSuite/pull/75 (will update the refs from my fork to a pinned commit in that repo after merging that PR).

## Overview

This adds a set of tests coming from generated test cases at https://github.com/nod-ai/SHARK-TestSuite/tree/main/iree_tests.

* We're not necessarily aiming for 100% of these tests passing, but we do want an automated way to verify compiler and runtime behavior across a wide range of cases.
* These new tests will run on presubmit, currently as part of the `pkgci` workflow. As larger tests are added we may segment the test suite into pre-submit/post-submit/nightly tests.
* I'm expecting that as new tests pass or fail we can update the XFAIL sets in each config file. As this test suite matures we may want to further filter tests by feature, platform, size, etc. -- that may need changes to the config format and pytest setup.

## Test suite architecture

Each test case is a directory with a few files following a naming scheme:

```
[program name 1]/
  model.mlir
  input_0.npy
  output_0.npy
  test_data_flags.txt
```

where `test_data_flags.txt` is a flagfile that tells `iree-run-module` about the inputs and expected outputs, e.g. `iree-run-module --flagfile=test_data_flags.txt` with these contents:
```
--input=@input_0.npy
--expected_output=@output_0.npy
```

The test cases themselves are not directly configured with any particular build system or test framework. Instead, a [pytest](https://docs.pytest.org/en/stable/) `conftest.py` script collects test files and defines test cases programmatically ([see pytest docs](https://docs.pytest.org/en/8.0.x/example/nonpython.html)).

Projects (or developers) that want to use the test suite must provide two things:

* The `iree-compile` and `iree-run-module` tools (e.g. from packages or a build/install directory), discoverable on `$PATH`
* A list of JSON configuration file paths in the `IREE_TEST_CONFIG_FILES` environment variable with contents like this:

    ```json
    {
      "config_name": "cpu",
      "iree_compile_flags": ["--iree-hal-target-backends=llvm-cpu"],
      "iree_run_module_flags": ["--device=local-task"],
      "skip_compile_tests": [],
      "skip_run_tests": [],
      "expected_compile_failures": ["test_abs"],
      "expected_run_failures": ["test_add"]
    }
    ```

The `conftest.py` script will load the configuration files and create test cases using the provided flags and XFAIL sets.

To run the tests:

```bash
$ export IREE_TEST_CONFIG_FILES=/path/to/config_1.json;/path/to/config_2.json
$ pytest SHARK-TestSuite/iree_tests
```

## Initial test cases

The test suite currently contains 1047 test cases converted from the ONNX test suite at https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node.

* On CPU, 700 test cases currently fail to compile
* On CPU, 109 test cases currently fail to run (e.g. outputs diverge from ONNX)
* Another 234 test cases failed to convert from ONNX using `iree-import-onnx` and are excluded from the suite for now

Each test case is typically very small and the entire suite runs in ~15 seconds on a CI machine with 96 worker threads.

More test cases are expected to be added over time. In particular, we are interested in adding full model tests for programs in the current focus set and I'd like to see JAX/StableHLO and TOSA tests if we can convert from existing test suites using a similar strategy.

## Sample logs

* Passing: https://github.com/openxla/iree/actions/runs/8071186394/job/22050540064#step:10:17
* Failing: https://github.com/openxla/iree/actions/runs/8070306776/job/22047644180#step:10:17
  * I turned off one 'compile' XFAIL and all 'run' XFAILs to show the output formats

ci-exactly: build_packages,regression_test_amdgpu_rocm,regression_test_amdgpu_vulkan,regression_test_cpu

### CODE REVIEW

**Review by benvanik — COMMENTED** [review]
Neat!

I really like how the test suite is external but we have control over the configuration and which tests are run - disabling a test doesn't require anything out-of-tree/integration/etc.

If the tests are really small (not doing big matmuls/convs/etc) then you could run the CPU ones with local-sync, which has less overhead per run (no need to spin up a thread pool) and less contention when running the tests in parallel (otherwise 96 tests running will use 96*(whatever thread count we choose) threads).

Unfortunate the XFAIL'ing requires moving the test lines out and into another nest - I suspect over time we'll end up with out of order/missing tests like we do today with the cmake ones. Other thing is tests in one config that are missing from others by omission - we've had that before in stablehlo/tosa where someone added/removed a test in one and it was gone forever as keeping the lists in sync was too hard. It may be worth thinking about how the main list of tests available in the suite could be a single source of truth and the config-specific JSON files would only contain the XFAILs.

I think this is a great start, but before we scale much further I think it'd be worth getting a single test availability list together. e.g. a single JSON with the lists of tests available then the JSON per config like you have with config-specific xfails.

File: .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml
Comment (ScottTodd) [inline .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml]: Bah. Compiler crashes for Vulkan and ROCm that aren't even interacting nicely with pytest:
* https://github.com/openxla/iree/actions/runs/8072456690/job/22054484578?pr=16589#step:9:457
* https://github.com/openxla/iree/actions/runs/8072261795/job/22053870186#step:9:29

even with changes removed? Huh?? something wrong in the env / with the multithreading? https://github.com/openxla/iree/actions/runs/8072456690/job/22054484855?pr=16589#step:6:42

File: experimental/regression_suite/external_test_suite/config_cpu.json
Comment (benvanik) [inline experimental/regression_suite/external_test_suite/config_cpu.json]: this is really `cpu_llvm_task` (vs `cpu_vmvx_inline` or `cpu_llvm_sync` etc)

File: .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml
Comment (benvanik) [reply .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml]: this is during compilation it looks like? that's really weird!
(96 parallel tests creating vulkan/rocm devices seems like it'd be a stress test, but the compiler should be fine)

**Comment by ScottTodd** [comment]
> Unfortunate the XFAIL'ing requires moving the test lines out and into another nest - I suspect over time we'll end up with out of order/missing tests like we do today with the cmake ones. Other thing is tests in one config that are missing from others by omission - we've had that before in stablehlo/tosa where someone added/removed a test in one and it was gone forever as keeping the lists in sync was too hard. It may be worth thinking about how the main list of tests available in the suite could be a single source of truth and the config-specific JSON files would only contain the XFAILs.
> 
> I think this is a great start, but before we scale much further I think it'd be worth getting a single test availability list together. e.g. a single JSON with the lists of tests available then the JSON per config like you have with config-specific xfails.

Right now there is no list of "all tests" - tests are just collected from a directory and then checked against the XFAIL lists. If a test starts compiling but still fails to run then it would need to be moved between XFAIL lists, yes. I did want to support an "include list" as well as the "exclude lists" though, for cases where you want to specifically run only a subset even if new tests are added. Could be useful to share such a list between config files.

I'd considered using Python as the "configuration language" (with all the power and complexity that brings). In that case we could have a common.py that each config could extend 🤔

File: .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml
Comment (ScottTodd) [reply .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml]: Yeah I turned off the ROCm tests while I wanted to investigate on my machine... but it looks like something still needs debugging even with Vulkan. I can turn down/off the parallelism on the GPU machines to see if that helps at least isolate what is failing.

**Comment by benvanik** [comment]
oh, you're right! I missed that - there are so many failures I just assumed that was a full test list 🤦 
this is great then :)
it does mean that new tests added to the suite will need to be versioned with the config in the repo when they are added failing - we'll want to only bump to a new version of the suite in pulls that also add the xfails.

**Review by benvanik — APPROVED** [review]

**Comment by ScottTodd** [comment]
> it does mean that new tests added to the suite will need to be versioned with the config in the repo when they are added failing - we'll want to only bump to a new version of the suite in pulls that also add the xfails.

Yep. I'll pin to a specific commit in the test suite repo and then we can bump deliberately with updated XFAIL sets for any newly added or changed test cases. Until we update the test suite version any failures will be coming directly from `iree-compile` or `iree-run-module`.

**Comment by ScottTodd** [comment]
I also have a GitHub Action in the test suite repo running the CPU tests using nightly IREE releases. We can use that for extra testing or as a way to iterate on the test suite itself without needing to work across repos.

**Comment by ScottTodd** [comment]
BTW, the test files are in the torch dialect, e.g.

```mlir
module {
  func.func @test_abs(%arg0: !torch.vtensor<[3,4,5],f32>) -> !torch.vtensor<[3,4,5],f32> attributes {torch.onnx_meta.ir_version = 7 : si64, torch.onnx_meta.opset_version = 13 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
    %0 = torch.operator "onnx.Abs"(%arg0) : (!torch.vtensor<[3,4,5],f32>) -> !torch.vtensor<[3,4,5],f32>
    return %0 : !torch.vtensor<[3,4,5],f32>
  }
}
```

So we may need to regenerate them with torch-mlir / (and possibly llvm?) updates. That does introduce a synchronization point where work may be needed across multiple repositories. If we trust the code, we could disable tests to unblock those sorts of updates... or we could use these tests to validate such updates.

File: .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml
Comment (ScottTodd) [reply .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml]: It was (up to) 32 Vulkan devices yeah... Dropped down to 4 and the tests still only took 30 seconds but didn't have any crashes. HMMMM... will monitor carefully after merge I guess.

Comment (ScottTodd) [reply .github/workflows/pkgci_regression_test_amdgpu_vulkan.yml]: Seems stable with 4 threads instead of 32.

**Comment by ScottTodd** [comment]
Okay, I've fixed the XFAIL lists for Vulkan and ROCm. I think this is ready to merge. Sorta tempted to sync past the current LLVM integrate since this could have a race with it. Or maybe we could have our first failure 😅 (not expecting anything to break... but it could happen)

**Comment by stellaraccident** [comment]
> BTW, the test files are in the torch dialect, e.g.
> 
> ```mlir
> module {
>   func.func @test_abs(%arg0: !torch.vtensor<[3,4,5],f32>) -> !torch.vtensor<[3,4,5],f32> attributes {torch.onnx_meta.ir_version = 7 : si64, torch.onnx_meta.opset_version = 13 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
>     %0 = torch.operator "onnx.Abs"(%arg0) : (!torch.vtensor<[3,4,5],f32>) -> !torch.vtensor<[3,4,5],f32>
>     return %0 : !torch.vtensor<[3,4,5],f32>
>   }
> }
> ```
> 
> So we may need to regenerate them with torch-mlir / (and possibly llvm?) updates. That does introduce a synchronization point where work may be needed across multiple repositories. If we trust the code, we could disable tests to unblock those sorts of updates... or we could use these tests to validate such updates.

It should be rare. torch-mlir is mostly insulated from MLIR format churn, and I don't remember the last time we accepted a patch that broke its ASM format.

**Pull request merged**
